axelrod/strategies/dbs.py

```diff
@@ -0,0 +1,354 @@
+        # for a condition j: 
+        # G[i] = 1 if cond j was True at turn i-1 and C has been played
+        # by the opponent; else G[i]=0
+        # F[i] = 1 if cond j was True at turn i-1; else G[i]=0
+        # initial hypothesized policy is TitForTat
+        self.history_by_cond[(C,C)]=([1],[1])
+        self.history_by_cond[(C,D)]=([1],[1])
+        self.history_by_cond[(D,C)]=([0],[1])
+        self.history_by_cond[(D,D)]=([0],[1])
+
+
+    def reset(self):
+        super().reset()
+        self.Rd = Policy.prob_policy(1,1,0,0)
+        self.Rc = Policy()
+        self.Pi = self.Rd   # policy used by MoveGen
+        self.violation_counts = Policy()
+        #self.reject_threshold = reject_threshold
+        #self.violation_threshold = violation_threshold
+        #self.promotion_threshold = promotion_threshold
+        #self.tree_depth = tree_depth
+        self.v = 0
+        #self.alpha = discount_factor
+        self.history_by_cond = {}
+        self.history_by_cond[(C,C)]=([1],[1])
+        self.history_by_cond[(C,D)]=([1],[1])
+        self.history_by_cond[(D,C)]=([0],[1])
+        self.history_by_cond[(D,D)]=([0],[1])
+
+
+
+    def should_promote(self,r_plus,promotion_threshold=3):
+        if r_plus[1]==C:
+            to_check = 0
+        elif r_plus[1]==D:
+            to_check = 1
+        k=1
+        count = 0
+        while (
+                k<len(self.history_by_cond[r_plus[0]][0])
+                and not(self.history_by_cond[r_plus[0]][0][1:][-k]==to_check and self.history_by_cond[r_plus[0]][1][1:][-k]==1)):
+            if (self.history_by_cond[r_plus[0]][1][1:][-k]==1):
+                count += 1
+            k+=1
+        if (count>=promotion_threshold):
+            return True
+        return False
+
+    def should_demote(self,r_minus,violation_threshold=4):
+        if (self.violation_counts[r_minus[0]]>=violation_threshold):
+            return True
+        return False
+
+
+    def update_history_by_cond(self,opponent_history):
+        two_moves_ago = (self.history[-2],opponent_history[-2])
+        for outcome,GF in self.history_by_cond.items():
+            G,F = GF
+            if outcome == two_moves_ago:
+                if opponent_history[-1]==C:
+                    G.append(1)
+                else:
+                    G.append(0)
+                F.append(1)
+            else:
+                G.append(0)
+                F.append(0)
+
+    def compute_prob_rule(self,outcome,alpha):
+        G = self.history_by_cond[outcome][0]
+        F = self.history_by_cond[outcome][1]
+        discounted_g = 0
+        discounted_f = 0
+        alpha_k = 1
+        for g,f in zip(G[::-1],F[::-1]):
+            discounted_g += alpha_k*g
+            discounted_f += alpha_k*f
+            alpha_k = alpha*alpha_k
+        p_cond = discounted_g/discounted_f
+        return p_cond
+
+    def strategy(self, opponent: Player) -> Action:
+        """This is the actual strategy"""
```

Comment:
Can we have a more informative doc string (or none if the strategy description covers it all)